bokeh/core/property/bases.py

```diff
@@ -178,7 +178,18 @@ def matches(self, new, old):
         if isinstance(new, np.ndarray) or isinstance(old, np.ndarray):
             return np.array_equal(new, old)
 
-        return new == old
+        # this handles the special but common case where there is a dict with numpy
+        # arrays as values (e.g. the .data property of a ColumnDataSource)
+        if isinstance(new, dict) and isinstance(old, dict):
+            return list(new.keys()) == list(old.keys()) and \
+                all(self.matches(*args) for args in zip(new.values(), old.values()))
```

Comment:
Is this safe? The ordering of dicts isn't guaranteed right, so the order of keys and values ``old`` and ``new`` isn't necessarily the same (although they probably are in some python versions and implementations).